# Analysis of Colby's timestamp data

1. Create a pandas series with the timestamp data

In [1]:
import pandas as pd

In [2]:
colby_data = pd.read_csv('colby.csv')
colby_data

Unnamed: 0                       0
0              0               timestamp
1      timestamp   Aug 4, 2019, 12:00 PM
2      timestamp    Aug 2, 2019, 8:53 PM
3      timestamp  Jul 22, 2019, 10:29 PM
4      timestamp  Jul 22, 2019, 10:28 PM
...          ...                     ...
23341  timestamp    Apr 9, 2023, 2:31 PM
23342  timestamp   Apr 9, 2023, 11:38 AM
23343  timestamp   Apr 9, 2023, 11:33 AM
23344  timestamp   Apr 9, 2023, 11:33 AM
23345  timestamp   Apr 9, 2023, 10:57 AM

[23346 rows x 2 columns]

In [3]:
timestamps = colby_data.iloc[1:,1]
timestamps = pd.to_datetime(timestamps)
timestamps

1       2019-08-04 12:00:00
2       2019-08-02 20:53:00
3       2019-07-22 22:29:00
4       2019-07-22 22:28:00
5       2019-07-08 19:18:00
                ...        
23341   2023-04-09 14:31:00
23342   2023-04-09 11:38:00
23343   2023-04-09 11:33:00
23344   2023-04-09 11:33:00
23345   2023-04-09 10:57:00
Name: 0, Length: 23345, dtype: datetime64[ns]

2. Create dataframe with times between timestamps

In [4]:
timestamps = timestamps.sort_values()
df = pd.DataFrame({
    'timestamps': timestamps,
    'diffs': timestamps.diff()
    })
df

timestamps            diffs
46    2018-03-11 20:26:00              NaT
45    2018-03-12 20:59:00  1 days 00:33:00
44    2018-05-28 13:09:00 76 days 16:10:00
43    2018-06-24 22:07:00 27 days 08:58:00
42    2018-07-29 20:53:00 34 days 22:46:00
...                   ...              ...
23250 2023-04-13 08:16:00  0 days 00:00:00
23249 2023-04-13 08:20:00  0 days 00:04:00
23248 2023-04-13 08:22:00  0 days 00:02:00
23247 2023-04-13 08:23:00  0 days 00:01:00
23246 2023-04-13 08:26:00  0 days 00:03:00

[23345 rows x 2 columns]

3. Find longest timestamp difference and average timestamp difference

In [7]:
max_diff = df['diffs'].max()
print('Longest time between timestamps:', max_diff)

Longest time between timestamps: 76 days 16:10:00


In [8]:
avg_diff = df['diffs'].mean()
print('Average time between timestamps:', avg_diff)

Average time between timestamps: 0 days 01:54:38.615490061


4. Find the day with the most timestamps

In [40]:
df['year'] = pd.DatetimeIndex(df['timestamps']).year
df['day'] = pd.DatetimeIndex(df['timestamps']).dayofyear

In [58]:
timestamps_per_day = df.pivot_table(index=['year', 'day'], values='timestamps', aggfunc='count')
timestamps_per_day


timestamps
year day            
2018 70            1
     71            1
     148           1
     175           1
     210           1
...              ...
2023 99           64
     100          33
     101          12
     102          26
     103          11

[975 rows x 1 columns]

In [69]:
max_timestamps = timestamps_per_day['timestamps'].max()
timestamps_per_day[timestamps_per_day.timestamps == max_timestamps]

timestamps
year day            
2022 302         168

In [72]:
df[(df.year == 2022) & (df.day == 302)]

timestamps           diffs  year  day
18702 2022-10-29 07:32:00 0 days 09:59:00  2022  302
18701 2022-10-29 07:32:00 0 days 00:00:00  2022  302
18700 2022-10-29 07:32:00 0 days 00:00:00  2022  302
18699 2022-10-29 07:32:00 0 days 00:00:00  2022  302
18698 2022-10-29 07:53:00 0 days 00:21:00  2022  302
...                   ...             ...   ...  ...
18939 2022-10-29 21:10:00 0 days 00:00:00  2022  302
18938 2022-10-29 21:10:00 0 days 00:00:00  2022  302
18937 2022-10-29 21:10:00 0 days 00:00:00  2022  302
18936 2022-10-29 22:18:00 0 days 01:08:00  2022  302
18935 2022-10-29 22:18:00 0 days 00:00:00  2022  302

[168 rows x 4 columns]

From above, we can see that the day with the most timestamps was October 29, 2022, with 302 timestamps.

5. Find the month with the most timestamps

In [75]:
df['month'] = pd.DatetimeIndex(df['timestamps']).month
df.head()

timestamps            diffs  year  day  month
46 2018-03-11 20:26:00              NaT  2018   70      3
45 2018-03-12 20:59:00  1 days 00:33:00  2018   71      3
44 2018-05-28 13:09:00 76 days 16:10:00  2018  148      5
43 2018-06-24 22:07:00 27 days 08:58:00  2018  175      6
42 2018-07-29 20:53:00 34 days 22:46:00  2018  210      7

In [79]:
timestamps_per_month = df.pivot_table(index=['year', 'month'], values='timestamps', aggfunc='count')
timestamps_per_month.head(10)


timestamps
year month            
2018 3               2
     5               1
     6               1
     7               1
     8               2
     10              5
     12              1
2019 1               9
     2               5
     3               7

In [83]:
max_timestamps_month = timestamps_per_month['timestamps'].max()
max_timestamps_month

1778

In [84]:
timestamps_per_month[timestamps_per_month.timestamps == max_timestamps_month]

timestamps
year month            
2021 10           1778

We can see that the month with the most timestamps was October 2021, with 1778 timestamps.

6. Predict the source of the data

My guess is that this data shows the timestamps of some sort of interaction on social media. We see that the average time between timestamps is under 2 hours, which seems to indicate that the action the timestamp is describing happens quite often and in likely in clusters. My prediction is that this is showing the times when Colby liked something on social media, probably Instagram.